### PR TITLE
Add Azure Content Safety Tool to Prompt Flow promptflow-tools package

### PR DESCRIPTION
**Pull request by asmeyyh**
# Description

Users can seamlessly integrate the Azure Content Safety Tool into their workflow by simply installing promptflow-tools. The Azure Content Safety Tool is a critical feature for analyzing textual content, helping us provide a more comprehensive content safety solution within Prompt Flow. This tool offers an effective way to detect and filter unsafe or inappropriate content, contributing to the legitimacy and security of our platform.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [x] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [X] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

e2e test passed
image

### CODE REVIEW

File: src/promptflow-tools/promptflow/tools/azure_content_safety.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+
+    def build_headers(self) -> Dict[str, str]:
+        """
+        Builds the headers for the Content Safety API request.
+
+        Returns:
+        - dict[str, str]: The headers for the Content Safety API request.
+        """
+        return {
+            "Ocp-Apim-Subscription-Key": self.subscription_key,
```
Comment (chenslucky) [inline src/promptflow-tools/promptflow/tools/azure_content_safety.py:182]: The Spell Check CI doesn't recognize the word 'Apim'. Please add this word to the ignore words section in .cspell.json.

**Review by chenslucky — COMMENTED** [review]
Refactor the pr title and descriptions. Add more information about your tool.

File: src/promptflow-tools/promptflow/tools/azure_content_safety.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+    switcher = {
+        TextCategorySensitivity.DISABLE: -1,
+        TextCategorySensitivity.LOW_SENSITIVITY: 6,
+        TextCategorySensitivity.MEDIUM_SENSITIVITY: 4,
+        TextCategorySensitivity.HIGH_SENSITIVITY: 2,
+    }
+    return switcher.get(sensitivity, f"Non-supported sensitivity: {sensitivity}")
+
+
+class MediaType(enum.Enum):
```
Comment (chenslucky) [inline src/promptflow-tools/promptflow/tools/azure_content_safety.py:87]: Use `Enum` is enough?

File: src/promptflow-tools/promptflow/tools/azure_content_safety.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
+import enum
```
Comment (chenslucky) [inline src/promptflow-tools/promptflow/tools/azure_content_safety.py:1]: no need?

File: src/promptflow-tools/tests/conftest.py [hunk 4]
```diff
@@ -92,7 +97,8 @@ def skip_if_no_api_key(request, mocker):
         conn_name = request.node.get_closest_marker('skip_if_no_api_key').args[0]
         connection = request.getfixturevalue(conn_name)
         # if dummy placeholder key, skip.
```
Comment (chenslucky) [inline src/promptflow-tools/tests/conftest.py:99]: revert changes in conftest.py. Please add these changes when you add tests about your tool.

File: src/promptflow-tools/promptflow/tools/__init__.py [hunk 5]
```diff
@@ -1,3 +1,4 @@
 from .aoai import AzureOpenAI  # noqa: F401
 from .openai import OpenAI  # noqa: F401
 from .serpapi import SerpAPI  # noqa: F401
+from .azure_content_safety import AzureContentSafety, analyze_text  # noqa: F401
```
Comment (chenslucky) [inline src/promptflow-tools/promptflow/tools/__init__.py]: Why add analyze_text to this import

File: src/promptflow-tools/promptflow/tools/azure_content_safety.py [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+
+        Returns:
+        - str: The URL for the Content Safety API.
+        """
+        if media_type == MediaType.Text:
+            return f"{self.endpoint}/contentsafety/text:analyze?api-version={self.api_version}"
+        elif media_type == MediaType.Image:
+            return f"{self.endpoint}/contentsafety/image:analyze?api-version={self.api_version}"
+        else:
+            raise ValueError(f"Invalid Media Type {media_type}")
```
Comment (chenslucky) [inline src/promptflow-tools/promptflow/tools/azure_content_safety.py:172]: Create your own specific exception for your tool like errors in https://github.com/microsoft/promptflow/blob/main/src/promptflow-tools/promptflow/tools/exception.py

Comment (chenslucky) [reply src/promptflow-tools/promptflow/tools/azure_content_safety.py:172]: And please test if you have an invalid type, if flow throws the correct exception in vscode.

File: src/promptflow-tools/promptflow/tools/azure_content_safety.py [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+        if (
+                "blocklistsMatchResults" in detection_result
+                and detection_result["blocklistsMatchResults"]
+                and len(detection_result["blocklistsMatchResults"]) > 0
+        ):
+            final_action = Action.Reject
+
+        print(final_action.name)
+        print(action_result)
+        return Decision(final_action, action_result)
```
Comment (chenslucky) [inline src/promptflow-tools/promptflow/tools/azure_content_safety.py:281]: create specific exception for error handing, please refine all exception in your tool

File: src/promptflow-tools/promptflow/tools/azure_content_safety.py [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+                final_action = action
+
+        if (
+                "blocklistsMatchResults" in detection_result
+                and detection_result["blocklistsMatchResults"]
+                and len(detection_result["blocklistsMatchResults"]) > 0
+        ):
+            final_action = Action.Reject
+
+        print(final_action.name)
```
Comment (chenslucky) [inline src/promptflow-tools/promptflow/tools/azure_content_safety.py:279]: Why add this print? Is it useful or just use tests in local?

File: src/promptflow-tools/promptflow/tools/azure_content_safety.py (hunk 2 above)
Comment (asmeyyh) [reply src/promptflow-tools/promptflow/tools/azure_content_safety.py:87]: This is align with sample code of Azure Content Safety GA version.

File: src/promptflow-tools/promptflow/tools/azure_content_safety.py (hunk 3 above)
Comment (asmeyyh) [reply src/promptflow-tools/promptflow/tools/azure_content_safety.py:1]: This is align with sample code of Azure Content Safety GA version.

File: src/promptflow-tools/promptflow/tools/azure_content_safety.py (hunk 8 above)
Comment (asmeyyh) [reply src/promptflow-tools/promptflow/tools/azure_content_safety.py:279]: Yes, this output is useful for user interaction.

File: src/promptflow-tools/promptflow/tools/azure_content_safety.py [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+    Sexual = 3
+    Violence = 4
+
+
+class Action(enum.Enum):
+    Accept = "Accept"
+    Reject = "Reject"
+
+
+class DetectionError(Exception):
```
Comment (chenslucky) [inline src/promptflow-tools/promptflow/tools/azure_content_safety.py:105]: Where use this DetectionError

File: src/promptflow-tools/promptflow/tools/exception.py [hunk 10]
```diff
@@ -155,3 +155,17 @@ class OpenSourceLLMKeyValidationError(ToolValidationError):
...
+
+class AzureContentSafetyInputValueError(UserErrorException):
+    """Base exception raised when the input type of Azure Content Safety is invalid."""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs, target=ErrorTarget.TOOL)
+
+
+class AzureContentSafetySystemError(SystemErrorException):
+    """Base exception raised when the input type of Azure Content Safety is invalid."""
```
Comment (chenslucky) [inline src/promptflow-tools/promptflow/tools/exception.py:168]: invalid comments

File: src/promptflow-tools/promptflow/tools/azure_content_safety.py (hunk 8 above)
Comment (chenslucky) [reply src/promptflow-tools/promptflow/tools/azure_content_safety.py:287]: If it is useful, could you print with some explains like f"Action name: {final_action.name}". maybe you can print clearer for all print info in your tool.

File: src/promptflow-tools/promptflow/tools/azure_content_safety.py (hunk 9 above)
Comment (asmeyyh) [reply src/promptflow-tools/promptflow/tools/azure_content_safety.py]: remove it

File: src/promptflow-tools/promptflow/tools/azure_content_safety.py (hunk 8 above)
Comment (asmeyyh) [reply src/promptflow-tools/promptflow/tools/azure_content_safety.py]: update

**Review by chenslucky — APPROVED** [review]

**Review by chjinche — APPROVED** [review]

**Pull request merged**